test/test_torch.py

```diff
@@ -2466,39 +2466,179 @@ def _test_rpow(self, cast):
+                lambda: tensor.pow_(pow))
+            self.assertRaisesRegex(
+                RuntimeError,
+                "Integers to negative integer powers are not allowed.",
+                lambda: torch.pow(tensor, pow))
+            self.assertRaisesRegex(
+                RuntimeError,
+                "Integers to negative integer powers are not allowed.",
+                lambda: torch.pow(tensor, pow, out=out))
+
+    @torchtest.for_all_device_types()
+    @unittest.skipIf(not TEST_NUMPY, 'Numpy not found')
+    def test_long_tensor_pow_floats(self, device):
+        ints = [0, 1, 23, 4567]
+        floats = [0.0, 1 / 3, 1 / 2, 1.0, 3 / 2, 2.0]
+
+        tensor = torch.tensor(ints, dtype=torch.int64, device=device)
+        nparr = np.array(ints, dtype=np.int64)
+
+        if device == 'cuda' and not TEST_WITH_ROCM:
+            with self.assertRaises(AssertionError):
+                # This is a check that pow CUDA implementation is
+                # incompatible with Numpy:
+                # pow CUDA  4 ^ 0.5 = 1
+                # numpy pow 4 ^ 0.5 = 2
+                # This check must be deleted after pow CUDA is fixed
+                for pow in floats:
+                    expected = np.power(nparr, pow).astype(np.int64)
+
+                    actual = tensor.pow(pow)
+                    self.assertEqual(expected, actual.cpu().numpy())
+
+                    actual = tensor.clone()
+                    actual2 = actual.pow_(pow)
+                    self.assertEqual(expected, actual.cpu().numpy())
+                    self.assertEqual(expected, actual2.cpu().numpy())
+
+                    actual = torch.pow(tensor, pow)
+                    self.assertEqual(expected, actual.cpu().numpy())
+
+                    actual2 = torch.pow(tensor, pow, out=actual)
+                    self.assertEqual(expected, actual.cpu().numpy())
+                    self.assertEqual(expected, actual2.cpu().numpy())
+        else:
+            for pow in floats:
+                expected = np.power(nparr, pow).astype(np.int64)
+
+                actual = tensor.pow(pow)
+                self.assertEqual(expected, actual.cpu().numpy())
+
+                actual = tensor.clone()
+                actual2 = actual.pow_(pow)
+                self.assertEqual(expected, actual.cpu().numpy())
+                self.assertEqual(expected, actual2.cpu().numpy())
+
+                actual = torch.pow(tensor, pow)
+                self.assertEqual(expected, actual.cpu().numpy())
+
+                actual2 = torch.pow(tensor, pow, out=actual)
+                self.assertEqual(expected, actual.cpu().numpy())
+                self.assertEqual(expected, actual2.cpu().numpy())
+
+    @torchtest.for_all_device_types()
+    @unittest.skipIf(not TEST_NUMPY, 'Numpy not found')
+    def test_float_scalar_pow_float_tensor(self, device):
+        def assertEqual(x, y):
+            self.assertEqual(np.nan_to_num(x), np.nan_to_num(y), allow_inf=True)
```

Comment:
because NaN != NaN, but in this test if actual is NaN and expected is NaN the test should pass

Reply:
self.assertEqual should support NaN == NaN. Here's an example where it does: https://github.com/pytorch/pytorch/blob/5254b12002caf44a4df0ec56f6aea83877dd82f7/test/test_torch.py#L785

Reply:
That's not true, it compares (x-y) to 1e-5, so NaN-Nan != 1e-5, and self.assertEqual(np.nan, np.nan) produces this:
```
  File "test/test_torch.py", line 2585, in assertEqual
    self.assertEqual(np.nan, np.nan)
  File "/data/users/pbelevich/pbelevich-pytorch/test/common_utils.py", line 705, in assertEqual
    super(TestCase, self).assertLessEqual(abs(x - y), prec, message)
AssertionError: nan not less than or equal to 1e-05 :
```

Reply:
Okay, I see the problem. You're using `self.assertEqual` to compare two np arrays. I'm not sure what the behavior of `self.assertEqual` on np arrays are.  

However, `self.assertEqual(tensor, other_tensor)` does check nan comparison. See the following code: https://github.com/pytorch/pytorch/blob/3a59a9b36c7e4ad291cd3b9d8e89402c4eaa566c/test/common_utils.py#L640 .

Instead of comparing numpy arrays using `self.assertEqual`, we should be comparing Tensors. Our implementation of `assertEqual` adds a lot of additional checks for Tensor, including the NaN equality check.

Reply:
That works! Thanks!